feat: add compatibility enable / disable for Windows

cdrage opened the pull request:
feat: add compatibility enable / disable for Windows

### What does this PR do?

* Adds functionality to the pre-existing "Docker Compatibility" button to
enable / disable Docker Compatibility on Windows
* Fixes tests from previous #1697 
* Implements "restart podman machine" function for future button for #1697 
* Small changes to extension.ts functionality (showing the information).

### Screenshot/screencast of this PR

<!-- Please include a screenshot or a screencast explaining what is doing this PR -->

![Screenshot 2023-03-22 at 11 32 38 AM](https://user-images.githubusercontent.com/6422176/227009178-dfac4a49-edf9-4c18-b0b5-9dd08f00f88c.png)


### What issues does this PR fix or reference?

<!-- Please include any related issue from Podman Desktop repository (or from another issue tracker).
-->

Closes part of https://github.com/containers/podman-desktop/issues/1447

### How to test this PR?

<!-- Please explain steps to reproduce -->

On Windows, click on the "Docker Compatibility" button and follow the
steps to enable / disable. This should be user friendly. Any comments please mention!

Signed-off-by: Charlie Drage <charlie@charliedrage.com>

cdrage commented on extensions/podman/src/compatibility-mode.ts:
```diff
@@ -118,11 +177,75 @@ export class DarwinSocketCompatibility extends SocketCompatibility {
+    const dockerDown = !(await this.isDockerRunning());
+    if (dockerDown) {
+      const result = await extensionApi.window.showInformationMessage(
+        'Start Docker before continuing. To disable Podman compatibility mode, Docker must be started to override the emulated socket.',
```
@benoitf thoughts on UI for this?

The steps to "disable" the docker compatibility is to:
1. Tell the user to start Docker (docker should automatically remap the socket)
2. Restart Podman (it'll detect that docker is active and not emulate the socket).

What are your thoughts on this part. It seems a bit odd.. but this is what's recommended for enabling compatibility for Windows.

cdrage commented on extensions/podman/src/lifecycle-context.ts:
```diff
@@ -0,0 +1,77 @@
+import type { LifecycleContext, Logger } from '@podman-desktop/api';
+
+// This file is the same as in the main repo (plugin/lifecycle-context.ts)
+// but added to the extension in order to implement logging for lifecycle context.
```
Needs to be added / copy-and-pasted over so we can use the logging functionality to call the extension + restart the machines. Which seems to be the most cleanest / efficient way rather than using `extension.ts` functions.

cdrage commented:
Tests add, confirmed working on Windows. Ready for review 💯

lstocchi reviewed (commented):
First, super cool!! Really useful feature.
The enable part works fine for me but the disable gets stuck.
I have my docker running but podman desktop does not see it. Am i missing something?
![docker-started](https://user-images.githubusercontent.com/49404737/227250578-9fba67e9-eb8c-4bef-ba3a-7089f1518615.gif)

lstocchi commented:
> It's because Podman Desktop doesn't detect your Docker instance (we are checking through providers, not if the exe is running).
> 
> Maybe I should change the text to something along the line of asking that Docker is running within Podman Desktop?

Idk honestly, for me the message is clear but it's like there is a problem with the discovery of docker (not related to this PR). I never noticed but if i stop docker, then podman, then start docker, podman-desktop doesn't refresh and i don't see docker running... i only see that podman is stopped

Other thing (even this can be seen as different issue) is that if i have my podman machine stopped (podman-machine-default) when i enabled compatibility another machine (different from the last i was using) gets started. So maybe this is not what i want.
In the image below you can see i was using the default but a random one (or is there a reason i'm not aware of??) was picked up and started.
![image](https://user-images.githubusercontent.com/49404737/227300615-bdc8d5ef-ec67-46a0-8d91-2ad489987f18.png)

benoitf commented:
@cdrage we are aware on the machine being active so I'm wondering why we ask the user to select the machine ?

cdrage commented:
> @cdrage we are aware on the machine being active so I'm wondering why we ask the user to select the machine ?
> 
> or it is only if they're all stopped

Edge case. Only if they are all stopped. 

If we are already aware of the machine being active, it'll restart normally.

jeffmaury reviewed (commented):
I had the following configuration:

- Podman running in compatibility mode
- Docker not started

Once I clicked on Docker compatibily, I have this dialog:
![image](https://user-images.githubusercontent.com/695993/227586006-f8e2f4bd-3ebc-42a0-ae05-2dfcabb7670f.png)

Seems to me that automatically is not clear to me because I feed once I will chose this option, I will always have Docker being reached on the Docker socket which cannot be guaranted

If I choose Disable, then I have this dialog:
![image](https://user-images.githubusercontent.com/695993/227586534-45ba491a-e96f-412f-96cf-c24a79bc4627.png)

So I started Docker Desktop but the Docker CLI is still being directed to Podman. If I click continue, then I enter an infinite loop

cdrage commented:
> 

Ah, good catch.

That's intentional for the loop, I should of renamed it to "Try again" as it'll just check to see if Docker is running before continuining.

I'll also remove the wording for "automatically" as that's not a good UX experience.

lstocchi reviewed (approved):
My only concern is still related to https://github.com/containers/podman-desktop/issues/1819 
For the rest LGTM

jeffmaury commented:
I had the following configuration:

- Docker running
- Podman machine running (thus compatibility mode is not enabled)

I stopped Docker and click on Docker compatibility and asked for restart but there is nothing in the UI that says that Docker compatibility mode is enabled (only the warning disappeared so if you don't concentrate your attention on the UI you may miss it).
Now I don't get how the disable works as we're told to start Docker but the Podman machine is still running thus I am still entering an endless loop

cdrage commented:
> 

Good point! I've gone ahead and removed the "loop" and added some code comments so that we reintroduce it once #1819 has been fixed.

jeffmaury commented:
Still KO for me:
I had

- Podman running in compatibility mode (Docker not running)

So once I clicked on the task bar item, and noticed that I need to start Docker, I started Docker but never saw that Podman is not running anymore in compatibility mode once Docker was started and clicked OK:

![compat](https://user-images.githubusercontent.com/695993/229138602-a0432e0a-6347-4036-8135-78320040e809.gif)

Second scenario:
- Docker running
- Podman running (no compatibility mode)

Once I did all the steps, the dashboard does not reflect the new state but the task bar item does.
![compat1](https://user-images.githubusercontent.com/695993/229140872-6e88d0c3-f897-4665-be8f-51b62f7fff46.gif)

cdrage commented:
> Still KO for me: I had
> 
> * Podman running in compatibility mode (Docker not running)
> 
> So once I clicked on the task bar item, and noticed that I need to start Docker, I started Docker but never saw that Podman is not running anymore in compatibility mode once Docker was started and clicked OK:
> 
> ![compat](https://user-images.githubusercontent.com/695993/229138602-a0432e0a-6347-4036-8135-78320040e809.gif) [ ![compat](https://user-images.githubusercontent.com/695993/229138602-a0432e0a-6347-4036-8135-78320040e809.gif) ](https://user-images.githubusercontent.com/695993/229138602-a0432e0a-6347-4036-8135-78320040e809.gif) [ ](https://user-images.githubusercontent.com/695993/229138602-a0432e0a-6347-4036-8135-78320040e809.gif)
> 
> Second scenario:
> 
> * Docker running
> * Podman running (no compatibility mode)
> 
> Once I did all the steps, the dashboard does not reflect the new state but the task bar item does. ![compat1](https://user-images.githubusercontent.com/695993/229140872-6e88d0c3-f897-4665-be8f-51b62f7fff46.gif) [ ![compat1](https://user-images.githubusercontent.com/695993/229140872-6e88d0c3-f897-4665-be8f-51b62f7fff46.gif) ](https://user-images.githubusercontent.com/695993/229140872-6e88d0c3-f897-4665-be8f-51b62f7fff46.gif) [ ](https://user-images.githubusercontent.com/695993/229140872-6e88d0c3-f897-4665-be8f-51b62f7fff46.gif)

So all of this is because of https://github.com/containers/podman-desktop/issues/1819 of Podman Desktop not picking up Docker correctly when it quits / reopens..

The prompts are intentional and unfortunatley the only solution is to possibly restart Podman Desktop?

WDYT we should do @benoitf should we have a button instead that says "Click here to restart Podman Desktop" until #1819 is fixed?

benoitf commented:
@cdrage the only solution is to restart Podman Desktop or just to refresh providers ?

cdrage commented:
> @cdrage the only solution is to restart Podman Desktop or just to refresh providers ?

Refresh providers. We can add a button for that instead? Add it to the "try again" loop.

jeffmaury commented:
Still not working for me for disabling: I don't understand what user needs to to, I started Docker Desktop but he will not get the socket as it is being hold by the Podman machine so am I forgetting something ?

cdrage commented:
> Still not working for me for disabling: I don't understand what user needs to to, I started Docker Desktop but he will not get the socket as it is being hold by the Podman machine so am I forgetting something ?

Did you start Docker Desktop and then restart Podman Desktop after? I added the message: "This may require a restart of Podman Desktop.". This is due to bug: https://github.com/containers/podman-desktop/issues/1819

jeffmaury commented:
> > Still not working for me for disabling: I don't understand what user needs to to, I started Docker Desktop but he will not get the socket as it is being hold by the Podman machine so am I forgetting something ?
> 
> Did you start Docker Desktop and then restart Podman Desktop after? I added the message: "This may require a restart of Podman Desktop.". This is due to bug: #1819

Then it needs to be more explicit as I understood 'may require restart' as you will tell me if restart is required (like in Windows installations)

cdrage commented:
> > > Still not working for me for disabling: I don't understand what user needs to to, I started Docker Desktop but he will not get the socket as it is being hold by the Podman machine so am I forgetting something ?
> > 
> > 
> > Did you start Docker Desktop and then restart Podman Desktop after? I added the message: "This may require a restart of Podman Desktop.". This is due to bug: #1819
> 
> Then it needs to be more explicit as I understood 'may require restart' as you will tell me if restart is required (like in Windows installations)

Confirmed.

I have remove the "may require restart" on enablement, as that's not needed.

I have added the following for disabling, that will tell the user to restart Podman Desktop:

```
    // Temporary, due to: https://github.com/containers/podman-desktop/issues/1819
    // Tell the user to restart Podman Desktop
    await extensionApi.window.showInformationMessage(
      'Podman Desktop must be restarted to apply the changes.',
      'OK',
    );
```

jeffmaury commented on extensions/podman/src/compatibility-mode.ts:
```diff
@@ -239,14 +239,18 @@ export class WindowsSocketCompatibility extends SocketCompatibility {
+        'Start Docker before continuing. To disable Podman compatibility mode, Docker must be started to override the emulated socket.',
         'OK',
       );
       return;
```
I still don't understand the sequence:

- return seems extra to me
- if user starts Docker Desktop then it won't have the socket as the Podman machine has it so restarting the Podman machine is useless

cdrage replied:
When Docker Desktop starts up it automatically overrides the docker socket and that's the only way to "re enable" the socket mode.

Yup, restarting Podman Machine is useless and that's why I removed the comment.

jeffmaury replied:
This is where I disagree: start Docker Desktop while the Podman machine is still running will not have the Docker socket taken by the Docker machine because it's already taken by the Podman machine: see my scenario where I start with no machines running then start the Podman one then start Docker Desktop, then stop all of them and starting Docker Desktop and Podman machine and this is where we reached the right state

https://user-images.githubusercontent.com/695993/230144640-67fd3d6c-da99-478d-839e-c95e886a0b10.mp4

cdrage replied:
Ah, that's why I was getting confused. 

That's different on my Windows 10 machine. 

Starting Docker Desktop automatically "re-registers" the socket and I have no issues with having podman machine already running.

Are you using WSL2 / anything specific to your machine?

jeffmaury replied:
Nothing specific WSL2 Podman machines

benoitf commented:
what should we do with this PR ? 2 months without new activity
